cluster-autoscaler/utils/drain/drain.go

```diff
@@ -0,0 +1,192 @@
+				// sophisticated than this
+				if err == nil && ds != nil {
+					// Otherwise, treat daemonset-managed pods as unmanaged since
+					// DaemonSet Controller currently ignores the unschedulable bit.
+					// FIXME(mml): Add link to the issue concerning a proper way to drain
+					// daemonset pods, probably using taints.
+					daemonsetPod = true
+				} else {
+					return []*api.Pod{}, fmt.Errorf("deamonset for %s/%s is not present, err: %v", pod.Namespace, pod.Name, err)
+				}
+			} else {
+				daemonsetPod = true
+			}
+		} else if refKind == "Job" {
+			if checkReferences {
+				job, err := client.ExtensionsClient.Jobs(sr.Reference.Namespace).Get(sr.Reference.Name)
+
+				// Assume the only reason for an error is because the Job is
+				// gone/missing, not for any other cause.  TODO(mml): something more
+				// sophisticated than this
+				if err == nil && job != nil {
+					replicated = true
+				} else {
+					return []*api.Pod{}, fmt.Errorf("job for %s/%s is not available: err: %v", pod.Namespace, pod.Name, err)
+				}
+			} else {
+				replicated = true
+			}
+		} else if refKind == "ReplicaSet" {
+			if checkReferences {
+				rs, err := client.ExtensionsClient.ReplicaSets(sr.Reference.Namespace).Get(sr.Reference.Name)
+
+				// Assume the only reason for an error is because the RS is
+				// gone/missing, not for any other cause.  TODO(mml): something more
+				// sophisticated than this
+				if err == nil && rs != nil {
+					if rs.Spec.Replicas < minReplica {
+						return []*api.Pod{}, fmt.Errorf("replication controller for %s/%s has too few replicas spec: %d min: %d",
+							pod.Namespace, pod.Name, rs.Spec.Replicas, minReplica)
+					} else {
+						replicated = true
+					}
+				} else {
+					return []*api.Pod{}, fmt.Errorf("replication controller for %s/%s is not available, err: %v", pod.Namespace, pod.Name, err)
+				}
+			} else {
+				replicated = true
+			}
+		}
+		if !daemonsetPod && !replicated {
+			return []*api.Pod{}, fmt.Errorf("%s/%s is not replicated", pod.Namespace, pod.Name)
+		}
+		if !daemonsetPod && pod.Namespace == "kube-system" && skipNodesWithSystemPods {
+			return []*api.Pod{}, fmt.Errorf("non-deamons set, non-mirrored, kube-system pod present: %s", pod.Name)
+		}
+		if !daemonsetPod && HasLocalStorage(pod) && skipNodesWithLocalStorage {
+			return []*api.Pod{}, fmt.Errorf("pod with local storage present: %s", pod.Name)
+		}
+
+		if !daemonsetPod {
+			pods = append(pods, pod)
+		}
+	}
+	return pods, nil
+}
+
+// CreatorRefKind returns the kind of the creator of the pod.
+func CreatorRefKind(pod *api.Pod) (string, error) {
+	sr, err := CreatorRef(pod)
+	if err != nil {
+		return "", err
+	}
+	if sr == nil {
+		return "", nil
+	}
+	return sr.Reference.Kind, nil
+}
+
+// CreatorRefKind returns the kind of the creator reference of the pod.
+func CreatorRef(pod *api.Pod) (*api.SerializedReference, error) {
+	creatorRef, found := pod.ObjectMeta.Annotations[controller.CreatedByAnnotation]
```

Comment:
For the purpose of this PR we want to be compatible with kubectl drain which doesn't use ownerreference.